ONSdigital/eq-survey-runner
# EQ 69 Harden Portal Integration

**What**
Updates the JWT decoder and hardens the token handling. The token now must obtain all the rules in: https://digitaleq.atlassian.net/browse/EQ-69. The biggest changes are the inclusion of iat and exp in the payload and the inclusion of kid in the header.

There is now a production mode environment variable (EQ_PRODUCTION), when set to true all requests to the survey runner must set the url parameter 'token'. This should be an JWE encrypted signed JWT token.

The number of unit tests related to this functionality has been increase to cover all test cases in the story. Additional Flask unit tests have also been added.

**How to test**
1. Check out this branch
2. Run the unit tests by using the ./scripts/run_tests.sh script. All unit tests should pass.
3. Enable production mode by setting the EQ_PRODUCTION environment variable to true. Also set the public and private key variables:
- `export EQ_RRM_PUBLIC_KEY=cat ./jwt-test-keys/rrm-public.pem`
- `export EQ_SR_PRIVATE_KEY=cat ./jwt-test-keys/sr-private.pem`
1. Check that all survey runner urls are rejected with a 401 response.
2. Add a valid JWE token to the url and check the URLs now respond with a 200
3. Disable production mode by setting the EQ_PRODUCTION environment variable to false.
4. Check that all survey runner urls return a 200 response without a token
5. Check that all survey runner urls also accept a token parameter and that the token is used. (A token can be JWE encrypted signed JWT, an unencrypted signed JWT or an unencrypted unsigned JWT)

**Who can review**
Anyone apart from @warrenbailey
